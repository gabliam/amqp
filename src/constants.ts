export const LIST_QUEUE_BEAN = `LIST_QUEUE_BEAN`;

export const TYPE = {
  RabbitController: 'RabbitControllerType'
};

export const METADATA_KEY = {
  RabbitController: '_RabbitController',
  RabbitHandler: '_RabbitHandler'
};

export const ERRORS_MSGS = {
  DUPLICATED_CONTROLLER_DECORATOR: `Cannot apply @RabbitController decorator multiple times.`
};
