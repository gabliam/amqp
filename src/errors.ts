export class AmqpTimeout extends Error {
  constructor(m: string) {
    super(m);

    // Set the prototype explicitly.
    Object.setPrototypeOf(this, AmqpTimeout.prototype);
  }
}
