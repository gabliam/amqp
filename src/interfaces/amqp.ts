import amqp = require('amqplib');

export interface QueueOptions {
  /**
   * if true, scopes the queue to the connection (defaults to false)
   */
  exclusive: boolean;

  /**
   * if true, the queue will survive broker restarts,
   * modulo the effects of exclusive and autoDelete;
   * this defaults to true if not supplied, unlike the others
   */
  durable: boolean;

  /**
   * if true, the queue will be deleted when the number of consumers drops to zero (defaults to false)
   */
  autoDelete: boolean;

  /**
   * RabbitMQ extensions can also be supplied as options.
   * These typically require non-standard x-* keys and values, sent in the arguments table; e.g., 'x-expires'.
   * When supplied in options, the x- prefix for the key is removed; e.g., 'expires'.
   * Values supplied in options will overwrite any analogous field you put in options.arguments.
   */
  arguments?: { [name: string]: string };
}

export interface QueueConfiguration {
  queueName: string;

  options: QueueOptions;
}

export type ConsumerHandler = (
  chan: amqp.Channel
) => (msg: Message) => void | Promise<void>;

export type Handler = (content: any) => any | Promise<any>;

export interface Controller {
  [k: string]: Handler;
}

export interface MessageFields {
  consumerTag: string;
  deliveryTag: number;
  redelivered: boolean;
  exchange: string;
  routingKey: string;
}

export interface Message {
  content: Buffer;

  properties: SendOptions;

  fields: MessageFields;
}

export interface ConsumeOptions {
  /**
   * a name which the server will use to distinguish message deliveries for the consumer;
   * mustn't be already in use on the channel
   * It's usually easier to omit this, in which case the server will create a random name and supply it in the reply.
   */
  consumerTag?: string;

  /**
   * in theory, if true then the broker won't deliver messages to the consumer if they were also published on this connection;
   * RabbitMQ doesn't implement it though, and will ignore it. Defaults to false.
   */
  noLocal?: boolean;

  /**
   * if true, the broker won't expect an acknowledgement of messages delivered to this consumer;
   * i.e., it will dequeue messages as soon as they've been sent down the wire.
   * Defaults to false (i.e., you will be expected to acknowledge messages).
   */
  noAck?: boolean;

  /**
   * if true, the broker won't let anyone else consume from this queue;
   * if there already is a consumer, there goes your channel
   * (so usually only useful if you've made a 'private' queue by letting the server choose its name).
   */
  exclusive?: boolean;

  /**
   *  gives a priority to the consumer; higher priority consumers get messages in preference to lower priority consumers.
   * See this RabbitMQ extension's documentation
   */
  priority?: number;

  /**
   * arbitrary arguments. Go to town.
   */
  arguments?: object;
}

export interface SendOptions {
  /**
   * if supplied, the message will be discarded from a queue once
   * it's been there longer than the given number of milliseconds.
   * In the specification this is a string; numbers supplied here will be coerced to strings for transit.
   */
  expiration?: string;

  /**
   * If supplied, RabbitMQ will compare it to the username supplied when opening the connection,
   * and reject messages for which it does not match.
   */
  userId?: string;

  /**
   * an array of routing keys as strings;
   * messages will be routed to these routing keys in addition to that given as the routingKey parameter.
   * A string will be implicitly treated as an array containing just that string.
   * This will override any value given for CC in the headers parameter. NB The property names CC and BCC are case-sensitive.
   */
  CC?: string | string[];

  /**
   * a priority for the message; ignored by versions of RabbitMQ older than 3.5.0,
   * or if the queue is not a priority queue (see maxPriority above).
   */
  priority?: number;

  /**
   * If truthy, the message will survive broker restarts provided it's in a queue that also survives restarts.
   * Corresponds to, and overrides, the property deliveryMode.
   */
  persistent?: boolean;

  /**
   * Either 1 or falsey, meaning non-persistent; or, 2 or truthy, meaning persistent.
   * That's just obscure though. Use the option persistent instead.
   */
  deliveryMode?: boolean | 1 | 2;

  /****************************************
   * Used by RabbitMQ but not sent on to consumers:
   ****************************************/

  /**
   * if true, the message will be returned if it is not routed to a queue (i.e., if there are no bindings that match its routing key).
   */
  mandatory?: boolean;

  /**
   * like CC, except that the value will not be sent in the message headers to consumers.
   */
  BCC?: string | string[];

  /****************************************
   * Not used by RabbitMQ and not sent to consumers:
   ****************************************/

  /**
   * in the specification, this instructs the server to return the message if it is not able to be sent immediately to a consumer.
   * No longer implemented in RabbitMQ, and if true, will provoke a channel error, so it's best to leave it out.
   * Ignored by RabbitMQ (but may be useful for applications)
   */
  immediate?: boolean;

  /**
   * a MIME type for the message content
   */
  contentType?: string;

  /**
   * a MIME encoding for the message content
   */
  contentEncoding?: string;

  /**
   * application specific headers to be carried along with the message content.
   * The value as sent may be augmented by extension-specific fields if they are given in the parameters, for example,
   * 'CC', since these are encoded as message headers; the supplied value won't be mutated
   */
  headers?: object;

  /**
   * usually used to match replies to requests, or similar
   */
  correlationId?: string;

  /**
   * often used to name a queue to which the receiving application must send replies,
   * in an RPC scenario (many libraries assume this pattern)
   */
  replyTo?: string;

  /**
   * arbitrary application-specific identifier for the message
   */
  messageId?: string;

  /**
   * a timestamp for the message
   */
  timestamp?: number;

  /**
   * an arbitrary application-specific type for the message
   */
  type?: string;

  /**
   * an arbitrary identifier for the originating application
   */
  appId?: string;
}

export interface RabbitHandlerOptions {
  consumeOptions?: ConsumeOptions;

  sendOptions?: SendOptions;

  sendOptionsError?: SendOptions;
}

export interface ConsumeConfig {
  queueName: string;
  handler: ConsumerHandler;
  options?: ConsumeOptions;
}
