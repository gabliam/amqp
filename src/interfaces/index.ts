export * from './amqp';
export * from './metadata';
