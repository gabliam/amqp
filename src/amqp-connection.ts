import amqp = require('amqplib');
import { Queue } from './queue';
import * as assert from 'assert';
import { interfaces } from '@gabliam/core';
import {
  ConsumerHandler,
  ConsumeOptions,
  SendOptions,
  Message,
  ConsumeConfig
} from './interfaces';
import * as uuid from 'uuid';
import * as PromiseB from 'bluebird';
import { AmqpTimeout } from './errors';

enum ConnectionState {
  stopped,

  running,

  starting,

  stopping
}

export class AmqpConnection {
  private connection: amqp.Connection;

  private channel: amqp.Channel;

  private state = ConnectionState.stopped;

  private consumerList: ConsumeConfig[] = [];

  constructor(
    private url: string,
    private queues: Queue[],
    private valueExtractor: interfaces.ValueExtractor
  ) {}

  async start() {
    if (this.state !== ConnectionState.stopped) {
      return;
    }

    this.state = ConnectionState.starting;
    this.connection = await amqp.connect(this.url);
    const ch = (this.channel = await this.connection.createChannel());
    for (const queue of this.queues) {
      await ch.assertQueue(queue.queueName, queue.queueOptions);
    }

    for (const { queueName, handler, options } of this.consumerList) {
      await ch.consume(queueName, handler, options);
    }

    this.state = ConnectionState.running;
  }

  addConsume(
    queue: string,
    handler: ConsumerHandler,
    options?: ConsumeOptions
  ) {
    const queueName = this.getQueueName(queue);
    assert(this.queueExist(queueName), `queue ${queueName} doesn't exist`);
    this.consumerList.push({ queueName, handler, options });
  }

  async sendToQueue(
    queue: string,
    content: Buffer,
    msg: Message,
    options?: SendOptions
  ) {
    const queueName = this.getQueueName(queue);
    await this.channel.sendToQueue(queueName, content, options);
    await this.channel.ack(msg);
  }

  async sendAndReceive<T = any>(
    queue: string,
    content: any,
    options: SendOptions = {},
    timeout: number = 5000
  ): Promise<T> {
    let onTimeout = false;
    let promise = new PromiseB<T>((resolve, reject) => {
      const queueName = this.getQueueName(queue);
      if (!options.correlationId) {
        options.correlationId = uuid();
      }
      const correlationId = options.correlationId;

      if (!options.replyTo) {
        options.replyTo = `amqpSendAndReceive${uuid()}`;
      }

      if (!options.expiration) {
        options.expiration = '' + timeout;
      }

      const replyTo = options.replyTo;
      const chan = this.channel;

      // create new Queue for get the response
      chan
        .assertQueue(replyTo, { exclusive: true, autoDelete: true })
        .then(() => {
          return chan.consume(replyTo, (msg: Message) => {
            if (!onTimeout && msg.properties.correlationId === correlationId) {
              resolve(JSON.parse(msg.content.toString()));
              chan.ack(msg);
            }
          });
        })
        .then(() => {
          chan.sendToQueue(
            queueName,
            new Buffer(JSON.stringify(content)),
            options
          );
        })
        .catch(err => reject(err));
    });

    if (timeout) {
      promise = promise.timeout(timeout).catch(PromiseB.TimeoutError, e => {
        onTimeout = true;
        throw new AmqpTimeout(e.message);
      });
    }

    return promise;
  }

  async stop() {
    if (this.state !== ConnectionState.running) {
      return;
    }
    this.state = ConnectionState.stopping;

    await this.channel.close();
    await this.connection.close();

    this.state = ConnectionState.stopped;
  }

  queueExist(queueName: string) {
    for (const queue of this.queues) {
      if (queue.queueName === queueName) {
        return true;
      }
    }
    return false;
  }

  getQueueName(queueName: string) {
    return this.valueExtractor(
      `application.amqp.queues.${queueName}.queueName`,
      this.valueExtractor(queueName, queueName)
    );
  }
}
