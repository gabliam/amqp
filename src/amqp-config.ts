import {
  PluginConfig,
  Bean,
  Value,
  inject,
  VALUE_EXTRACTOR,
  interfaces
} from '@gabliam/core';
import { QueueConfiguration } from './interfaces';
import { schemaPlugin } from './schema';
import { Queue } from './queue';
import { AmqpConnection } from './amqp-connection';
import * as Joi from 'joi';

@PluginConfig()
export class AmqpConfig {
  @Value('application.amqp.queues', schemaPlugin)
  private queueConfig: { [k: string]: QueueConfiguration };

  @Value('application.amqp.url', Joi.string().required())
  private url: string;

  constructor(
    @inject(VALUE_EXTRACTOR) public valueExtractor: interfaces.ValueExtractor
  ) {}

  createQueue() {
    return Object.keys(this.queueConfig).map<Queue>((k: string) => {
      const q = this.queueConfig[k];
      return new Queue(q.queueName, q.options);
    });
  }

  @Bean(AmqpConnection)
  async createConnection() {
    const connection = new AmqpConnection(
      this.url,
      this.createQueue(),
      this.valueExtractor
    );
    return connection;
  }
}
