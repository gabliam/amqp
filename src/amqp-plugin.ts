import amqp = require('amqplib');
import {
  interfaces as coreInterfaces,
  Scan,
  inversifyInterfaces,
  Registry,
  Plugin
} from '@gabliam/core';
import { TYPE, METADATA_KEY } from './constants';
import {
  RabbitHandlerMetadata,
  ConsumerHandler,
  Message,
  SendOptions,
  Controller
} from './interfaces';
import { AmqpConnection } from './amqp-connection';
import * as d from 'debug';

const debug = d('Gabliam:Plugin:AmqpPlugin');

@Plugin()
@Scan()
export class AmqpPlugin implements coreInterfaces.GabliamPlugin {
  /**
   * binding phase
   *
   * Bind all controller and bind express app
   * @param  {inversifyInterfaces.Container} container
   * @param  {Registry} registry
   */
  bind(container: inversifyInterfaces.Container, registry: Registry) {
    debug('bind AmqpPlugin');
    registry.get(TYPE.RabbitController).forEach(({ id, target }) =>
      container
        .bind(id)
        .to(target)
        .inSingletonScope()
    );
  }

  build(container: inversifyInterfaces.Container, registry: Registry) {
    debug('build AmqpPlugin');
    const controllerIds = registry.get(TYPE.RabbitController);
    const connection = container.get(AmqpConnection);
    debug('controllerIds', controllerIds, container);
    controllerIds.forEach(({ id: controllerId }) => {
      const controller = container.get<any>(controllerId);
      debug('controller', controller);

      const handlerMetadatas: RabbitHandlerMetadata[] = Reflect.getOwnMetadata(
        METADATA_KEY.RabbitHandler,
        controller.constructor
      );

      if (handlerMetadatas) {
        handlerMetadatas.forEach(handlerMetadata => {
          let consumeHandler: ConsumerHandler;
          if (handlerMetadata.type === 'Listener') {
            consumeHandler = this.constructListener(
              handlerMetadata,
              container,
              controllerId
            );
          } else {
            consumeHandler = this.constructConsumer(
              handlerMetadata,
              container,
              controllerId
            );
          }

          connection.addConsume(
            handlerMetadata.queue,
            consumeHandler,
            handlerMetadata.consumeOptions
          );
        });
      }
    });
  }

  async start(container: inversifyInterfaces.Container, registry: Registry) {
    const connection = container.get(AmqpConnection);
    await connection.start();
  }

  async stop(container: inversifyInterfaces.Container, registry: Registry) {
    const connection = container.get(AmqpConnection);
    await connection.stop();
  }

  private constructListener(
    handlerMetadata: RabbitHandlerMetadata,
    container: inversifyInterfaces.Container,
    controllerId: any
  ): ConsumerHandler {
    return (chan: amqp.Channel) => async (msg: Message) => {
      const connection = container.get(AmqpConnection);
      const controller = container.get<Controller>(controllerId);
      const content = connection.parseContent(msg);
      await Promise.resolve(controller[handlerMetadata.key](content));
      await chan.ack(msg);
    };
  }

  private constructConsumer(
    handlerMetadata: RabbitHandlerMetadata,
    container: inversifyInterfaces.Container,
    controllerId: any
  ): ConsumerHandler {
    return (chan: amqp.Channel) => async (msg: Message) => {
      // catch when error amqp (untestable)
      /* istanbul ignore next */
      if (msg.properties.replyTo === undefined) {
        throw new Error(`replyTo is missing`);
      }

      const connection = container.get(AmqpConnection);
      const controller = container.get<Controller>(controllerId);
      const content = connection.parseContent(msg);

      let response: any;
      let sendOptions: SendOptions;
      try {
        response = await Promise.resolve(
          controller[handlerMetadata.key](content)
        );
        sendOptions = handlerMetadata.sendOptions || {};
      } catch (err) {
        response = err;
        sendOptions = handlerMetadata.sendOptionsError || {};
      }

      connection.sendToQueueAck(msg.properties.replyTo, response, msg, {
        correlationId: msg.properties.correlationId,
        contentType: 'application/json',
        ...sendOptions
      });
    };
  }
}
