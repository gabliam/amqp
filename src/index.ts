export * from './amqp-connection';
export * from './interfaces';
export * from './decorators';
export * from './queue';
export * from './errors';

import { AmqpPlugin } from './amqp-plugin';

export { AmqpPlugin as default };
