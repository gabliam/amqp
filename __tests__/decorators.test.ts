import { RegistryMetada } from '@gabliam/core/lib/interfaces';
import { METADATA_KEY as CORE_METADATA_KEY } from '@gabliam/core/lib/constants';
import { METADATA_KEY } from '../src/constants';
import {
  RabbitController,
  RabbitListener,
  RabbitHandlerMetadata,
  RabbitConsumer
} from '../src/index';

test(`@RabbitController() decorator`, () => {
  @RabbitController()
  class RabbitControllerTest {}

  const entityMetadata: RegistryMetada = Reflect.getMetadata(
    CORE_METADATA_KEY.register,
    RabbitControllerTest
  );

  const documentMetadata: boolean = Reflect.getOwnMetadata(
    METADATA_KEY.RabbitController,
    RabbitControllerTest
  );

  expect(entityMetadata).toMatchSnapshot();
  expect(documentMetadata).toMatchSnapshot();
});

test('should fail when decorated multiple times the same class with @RabbitController', () => {
  expect(() => {
    @RabbitController()
    @RabbitController()
    class RabbitControllerTest {}
    // tslint:disable-next-line:no-unused-expression
    new RabbitControllerTest();
  }).toThrowError();
});

test(`@RabbitListener() decorator`, () => {
  @RabbitController()
  class RabbitControllerTest {
    @RabbitListener('listenerTest')
    listener() {}
  }

  const entityMetadata: RegistryMetada = Reflect.getMetadata(
    CORE_METADATA_KEY.register,
    RabbitControllerTest
  );

  const documentMetadata: boolean = Reflect.getOwnMetadata(
    METADATA_KEY.RabbitController,
    RabbitControllerTest
  );

  const handlerMetadatas: RabbitHandlerMetadata[] = Reflect.getOwnMetadata(
    METADATA_KEY.RabbitHandler,
    RabbitControllerTest
  );

  expect(entityMetadata).toMatchSnapshot();
  expect(documentMetadata).toMatchSnapshot();
  expect(handlerMetadatas).toMatchSnapshot();
});

test(`@RabbitConsumer() decorator`, () => {
  @RabbitController()
  class RabbitControllerTest {
    @RabbitConsumer('consumerTest')
    consumer() {}
  }

  const entityMetadata: RegistryMetada = Reflect.getMetadata(
    CORE_METADATA_KEY.register,
    RabbitControllerTest
  );

  const documentMetadata: boolean = Reflect.getOwnMetadata(
    METADATA_KEY.RabbitController,
    RabbitControllerTest
  );

  const handlerMetadatas: RabbitHandlerMetadata[] = Reflect.getOwnMetadata(
    METADATA_KEY.RabbitHandler,
    RabbitControllerTest
  );

  expect(entityMetadata).toMatchSnapshot();
  expect(documentMetadata).toMatchSnapshot();
  expect(handlerMetadatas).toMatchSnapshot();
});

test(`@RabbitListener() & @RabbitConsumer() decorator`, () => {
  @RabbitController()
  class RabbitControllerTest {
    @RabbitListener('listenerTest')
    listener() {}

    @RabbitConsumer('consumerTest')
    consumer() {}
  }

  const entityMetadata: RegistryMetada = Reflect.getMetadata(
    CORE_METADATA_KEY.register,
    RabbitControllerTest
  );

  const documentMetadata: boolean = Reflect.getOwnMetadata(
    METADATA_KEY.RabbitController,
    RabbitControllerTest
  );

  const handlerMetadatas: RabbitHandlerMetadata[] = Reflect.getOwnMetadata(
    METADATA_KEY.RabbitHandler,
    RabbitControllerTest
  );

  expect(entityMetadata).toMatchSnapshot();
  expect(documentMetadata).toMatchSnapshot();
  expect(handlerMetadatas).toMatchSnapshot();
});
