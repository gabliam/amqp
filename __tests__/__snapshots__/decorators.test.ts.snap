// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@RabbitConsumer() decorator 1`] = `
Object {
  "type": "RabbitControllerType",
  "value": Object {
    "id": [Function],
    "target": [Function],
  },
}
`;

exports[`@RabbitConsumer() decorator 2`] = `true`;

exports[`@RabbitConsumer() decorator 3`] = `
Array [
  Object {
    "key": "consumer",
    "queue": "consumerTest",
    "type": "Consumer",
  },
]
`;

exports[`@RabbitController() decorator 1`] = `
Object {
  "type": "RabbitControllerType",
  "value": Object {
    "id": [Function],
    "target": [Function],
  },
}
`;

exports[`@RabbitController() decorator 2`] = `true`;

exports[`@RabbitListener() & @RabbitConsumer() decorator 1`] = `
Object {
  "type": "RabbitControllerType",
  "value": Object {
    "id": [Function],
    "target": [Function],
  },
}
`;

exports[`@RabbitListener() & @RabbitConsumer() decorator 2`] = `true`;

exports[`@RabbitListener() & @RabbitConsumer() decorator 3`] = `
Array [
  Object {
    "key": "listener",
    "queue": "listenerTest",
    "type": "Listener",
  },
  Object {
    "key": "consumer",
    "queue": "consumerTest",
    "type": "Consumer",
  },
]
`;

exports[`@RabbitListener() decorator 1`] = `
Object {
  "type": "RabbitControllerType",
  "value": Object {
    "id": [Function],
    "target": [Function],
  },
}
`;

exports[`@RabbitListener() decorator 2`] = `true`;

exports[`@RabbitListener() decorator 3`] = `
Array [
  Object {
    "key": "listener",
    "queue": "listenerTest",
    "type": "Listener",
  },
]
`;
